crates/fuel-core/src/graphql_api/validation_extension.rs

```diff
@@ -0,0 +1,109 @@
+use crate::{
+    fuel_core_graphql_api::validation_extension::visitor::{
+        visit,
+        RuleError,
+        VisitorContext,
+    },
+    graphql_api::validation_extension::recursion_finder::RecursionFinder,
+};
+use async_graphql::{
+    extensions::{
+        Extension,
+        ExtensionContext,
+        ExtensionFactory,
+        NextParseQuery,
+        NextValidation,
+    },
+    parser::types::ExecutableDocument,
+    ServerError,
+    ServerResult,
+    ValidationResult,
+    Variables,
+};
+use std::sync::{
+    Arc,
+    Mutex,
+};
+
+mod recursion_finder;
+mod visitor;
+
+/// The extension validates the queries.
+pub(crate) struct ValidationExtension {
+    recursion_limit: usize,
+}
+
+impl ValidationExtension {
+    pub fn new(recursion_limit: usize) -> Self {
+        Self { recursion_limit }
+    }
+}
+
+impl ExtensionFactory for ValidationExtension {
+    fn create(&self) -> Arc<dyn Extension> {
+        Arc::new(ValidationInner::new(self.recursion_limit))
+    }
+}
+
+struct ValidationInner {
+    recursion_limit: usize,
+    errors: Mutex<Option<Vec<RuleError>>>,
```

Comment:
Do we need the option here? I.e. is it meaningful to differentiate between an empty vec of errors and no errors? Otherwise I think we can simplify this to
```suggestion
    errors: Mutex<Vec<RuleError>>,
```